tests/ga/test_firewall_manager.py

```diff
@@ -0,0 +1,155 @@
+
+    def _test_remove_legacy_rule_should_delete_the_legacy_rule(self, firewall_cmd_type, mock_firewall_cmd_type):
+        with mock_firewall_cmd_type() as mock:
+            firewall = firewall_cmd_type('168.63.129.16')
+            firewall.remove_legacy_rule()
+
+            self.assertEqual(
+                [
+                    mock.get_legacy_command(mock.check_option),
+                    mock.get_legacy_command(mock.delete_option)
+                ],
+                mock.call_list,
+                "Expected a check ({0}) for the legacy rule, followed by a delete ({1}) of the rule".format(mock.check_option, mock.delete_option))
+
+
+class TestIpTables(_TestFirewallCommand):
+    def test_setup_should_set_all_the_firewall_rules(self):
+        self._test_setup_should_set_all_the_firewall_rules(IpTables, MockIpTables)
+
+    def test_remove_should_delete_all_rules(self):
+        self._test_remove_should_delete_all_rules(IpTables, MockIpTables)
+
+    def test_remove_should_not_attempt_to_delete_rules_that_do_not_exist(self):
+        self._test_remove_should_not_attempt_to_delete_rules_that_do_not_exist(IpTables, MockIpTables)
+
+    def test_check_should_verify_all_rules(self):
+        self._test_check_should_verify_all_rules(IpTables, MockIpTables)
+
+    def test_remove_legacy_rule_should_delete_the_legacy_rule(self):
+        self._test_remove_legacy_rule_should_delete_the_legacy_rule(IpTables, MockIpTables)
+
+    def test_it_should_not_use_the_wait_option_on_iptables_versions_less_than_1_4_21(self):
+        with MockIpTables(version='1.4.20') as mock_iptables:
+            firewall = IpTables('168.63.129.16')
+            firewall.setup()
+
+            self.assertEqual(
+                [
+                    MockIpTables.get_accept_dns_command("-A").replace("-w ", ""),
+                    MockIpTables.get_accept_command("-A").replace("-w ", ""),
+                    MockIpTables.get_drop_command("-A").replace("-w ", "")
+                ],
+                mock_iptables.call_list,
+                "Expected only 3 calls to the -A (append) command without the -w option")
+
+
+class TestFirewallCmd(_TestFirewallCommand):
+    def test_setup_should_set_all_the_firewall_rules(self):
+        self._test_setup_should_set_all_the_firewall_rules(FirewallCmd, MockFirewallCmd)
+
+    def test_remove_should_delete_all_rules(self):
+        self._test_remove_should_delete_all_rules(FirewallCmd, MockFirewallCmd)
+
+    def test_remove_should_not_attempt_to_delete_rules_that_do_not_exist(self):
+        self._test_remove_should_not_attempt_to_delete_rules_that_do_not_exist(FirewallCmd, MockFirewallCmd)
+
+    def test_check_should_verify_all_rules(self):
+        self._test_check_should_verify_all_rules(FirewallCmd, MockFirewallCmd)
+
+    def test_remove_legacy_rule_should_delete_the_legacy_rule(self):
+        self._test_remove_legacy_rule_should_delete_the_legacy_rule(FirewallCmd, MockFirewallCmd)
+
```

Comment:
do you plan to add test cases for ntftbales later?

Reply:
I had to add unit tests for the issue you mentioned above, wrt ntf not raising an exception when the command is not available. As part of the fix for that, I added all unit test I had planned for nftables.

Unit tests for the env thread coming soon in a separate PR